### ISSUE TITLE
Rules and structure to get your assets in the Assets Store

### ISSUE THREAD

**Issue by Entr0py404** [4ian/GDevelop#2146]
# A [community managed Trello board is available here](https://trello.com/b/xoOCKFOf/gdevelop-assets) to submit assets to the asset store.

Guidelines and acceptable content are explained in the link you can find in the next message 👇

**Comment by 4ian** [comment]
Hey @Entr0py404, I'll publish soon here and probably on the wiki a list of guidelines to organize/name images in folders so that they are ready to be imported in the Assets Store :)

**Comment by 4ian** [comment]
Here are the rules for the assets to be imported in the assets store :)

The main idea is that assets are grouped into folders and sub folders. Images, audio and fonts should have a name that is the plain English name that will be displayed to the user. By giving specific names to images, they will be used to create a single object.

> Note that tilesheets **are not supported**! Each image of an object must **be in a separate file** (using the rules to name the files below). It's also recommended that all images of a single object have the same size (with transparent margins), because for now we can't specify the origin of an image.

### For images (to make sprite, tiled sprite or panel sprite objects):

Each image will be used to create an object. For sprites, more than one image can be used to create a single object, by following the rules described below.

- By default, an image will be used to create a **Sprite** object.
  - The **name of the file**, without the extension, will be used for the object name.

    > For example, an image called "Spaceship.png" will create a Sprite object called "Spaceship", with this image as the first animation.

  - If the object is **animated**, use different image files with the **same base name**, and add an underscore followed by the image index: `Spaceship_1.png`, `Spaceship_2.png`, `Spaceship_3.png`.


  - If the object has **multiple animations**, you can add the name of the animation between the *base name* and the index.

    > For example: `Spaceship_flying_1.png`, `Spaceship_flying_2.png`, `Spaceship_flying_3.png`, `Spaceship_1.png`, `Spaceship_idle.png`, `Spaceship_damaged_1.png`, `Spaceship_damaged_2.png` will create an object with three animations: Flying (3 images), Idle (1 image), Damaged (2 images)

  - To specify the **speed of the animations and if they are looping**, create a file with the **same base name** suffixed by `_METADATA.json`. For example: `Spaceship_METADATA.json`. Inside, put this JSON:

    ```json
    {
        "timeBetweenFrames": 0.2,
        "loop": false
    }
    ```
    
    > For example: ![image](https://user-images.githubusercontent.com/1280130/103412855-78836300-4b77-11eb-99ec-ac8287372ece.png)
    
    - In the future, we'll support one `_METADATA.json` file for each animation. For now, just one is supported for all animations of the object.

- If you want to create a **Tiled sprite** (for backgrounds or repeating patterns):
  - Prefix the *base name* by `tiled_`. For example: `tiled_Grass Background.png`

- If your image is a **9 patch**, that can be stretched with the center and the border repeated:
  - Slice your image in 9 different images, and name them: `9patch_base name_gravity.png` with gravity being: `southwest`, `southeast`, `south`, `northwest`, `northeast`, `north`, `west` and `east`.

  > For example: 9patch_Orange leaves_center.png, 9patch_Orange leaves_east.png, 9patch_Orange leaves_north.png, 9patch_Orange leaves_northeast.png, 9patch_Orange leaves_northwest.png, 9patch_Orange leaves_south.png, 9patch_Orange leaves_southeast.png, 9patch_Orange leaves_southwest.png, 9patch_Orange leaves_west.png,.

### Organizing the assets with tags

Put all your images in a folder. In this folder, create a file called `TAGS.md`.
In this file, put the tags of the assets, separated by a comma. Put at least a tag `top-down`, `side view` or `interface`. For example:

```
side view, pirate, ship, sea
```

You can put your images in different sub folders. Each sub folder will be used as a tag name. Some folders names like "PNG", "SVG", etc... will be ignored (so don't change the whole structure of your assets, we'll ignore anything that does not make sense).

Note that you can also put assets in a folder called `Unimplemented` or `TODO`, and these folders will be entirely ignored.

## For audio

Put your .aac or .wav files (prefer .aac) directly in a folder, with the name to be displayed: `Laser effect.aac`, `Background music.aac`.

`TAGS.md` files will work like for images. Sub folders are also used for tags as described for images.

## For fonts

Like images and audio, put your fonts in folders. `.ttf` and `.otf` files are supported.

`TAGS.md` files will work like for images. Sub folders are also used for tags as described for images.

## (⚠️ Advanced!) To make an object from scratch, for example particle emitters

> These assets from scratch are a bit cumbersome to author right now. Here is an example folder that shows how the particle emitter objects from Wishforge Games are done: [Particles.zip](https://github.com/4ian/GDevelop/files/5757421/Particles.zip)

Create an file with extension `.asset.json`, for example: `My Particle Effect.asset.json`. The format of this asset is:

```json
{
  "id": "",
  "name": "",
  "authors": [""],
  "license": "",
  "shortDescription": "",
  "description": "Enter a description. Leave the rest empty, it will be autogenerated",
  "previewImageUrls": [],
  "tags": [],
  "objectAssets": [
    {
      "object": { ...JSON CONTENT OF AN OBJECT...  },
      "resources": [ ],
      "customization": []
    }
  ],
  "gdevelopVersion": "",
  "version": ""
}
```

Leave almost everything empty. Create your object in GDevelop, then copy it in the clipboard. Paste the content in a file, and then copy just the "content.object" key:
![image](https://user-images.githubusercontent.com/1280130/103413204-d5334d80-4b78-11eb-8248-aecc67d17cba.png)

Paste it inside the `.asset.json` file you created, replacing `...JSON CONTENT OF AN OBJECT...`: 
![image](https://user-images.githubusercontent.com/1280130/103413230-f300b280-4b78-11eb-9c3b-718bdf58ba66.png)

Finally, if your object uses image resources, make sure to add them in the `resources` array:
![image](https://user-images.githubusercontent.com/1280130/103413269-1e839d00-4b79-11eb-91a6-39bc80696a98.png)

Put all images with the name that you put in the `resources` array next to your `.asset.json` file.

Finally, add:
- a file with the same name but with the extension `.preview.png` that will be shown in the assets store
- an empty filed called `IGNORED_FOR_IMAGE_ASSETS.md` so that the importer will ignore the images and won't try to create objects automatically :) 

> For example: ![image](https://user-images.githubusercontent.com/1280130/103413361-74f0db80-4b79-11eb-94c2-7cf535bb9118.png)


## License

Put a license.txt file in your folder. The importing will recognize it and apply to all files in the folder :)
Make sure to put your name in this license file too (or the name of the original author). 
The importer will be adapted to understand the license file.

## Something unclear?

Just ask me in this thread :) If the rules here are too complicated or are giving you too much work, you can suggest me a new rule and I'll see if I can adapt the importer so that it can understand your asset structure :) 
The goal is to make objects more or less automatically!

**Comment by tristanbob** [comment]
How are we going to deal with file-name collisions?  I am working on a large project now and have experienced the frustrating problem when you have two objects that use the same file name.

**Comment by arthuro555** [comment]
> (⚠️ Advanced!) To make an object from scratch, for example particle emitters

Maybe we could add an option to export an object to the asset store format from the IDE in the future?

**Comment by 4ian** [comment]
> How are we going to deal with file-name collisions? I am working on a large project now and have experienced the frustrating problem when you have two objects that use the same file name.

It's handled by GDevelop itself. When an asset is added, GDevelop looks for all resources:
- is this a resource (i.e: an image/font/audio file) that was already imported from the asset store? If yes, don't download it again.
- otherwise, then it's downlaoded and if there is a resource already existing with this name, a new name is generated (basically a number is added at the end of the resource and file name: 2, 3, 4...). Of course, this means the imported object will also get its internal resource names updated.

In other words: no worries :) If there are 100 objects in 100 different folders that are all named "Player.png" (generating an object called Player), you'll still be able to search and import all of them without clashes - and without clashes with your own existing resources.

> Maybe we could add an option to export an object to the asset store format from the IDE in the future?

Totally. By lack of time I've not added it already, but if someone wants to do that, it should be doable to have an option that would export the object in a `.asset.json` file, with the resource (images basically) next to it.

**Comment by tristanbob** [comment]
> It's handled by GDevelop itself. When an asset is added, GDevelop looks for all resources:

Does GDevelop beta102 currently handle name collisions when importing assets?  This is the second time I've run into this problem on this project.

![Screenshot 2020-12-31 170957](https://user-images.githubusercontent.com/8879811/103431300-16345d80-4b8b-11eb-8cb6-d037ded21824.jpg)

![Screenshot 2020-12-31 165936](https://user-images.githubusercontent.com/8879811/103431301-16ccf400-4b8b-11eb-9d41-753384b404b5.jpg)

**Comment by Entr0py404** [comment]
I'm trying to make it a bit easier for my self to make the files for the time being but is this the right format? 
https://www.youtube.com/watch?v=XfVbTVCW8rs

where is the description shown in the UI of GDev?, I think I'm getting confused by the automatically generated ones.

**Comment by 4ian** [comment]
> I'm trying to make it a bit easier for my self to make the files for the time being but is this the right format?

This looks mostly correct!
Only thing I could spot is that the resource files in the json are have a relative path with a sub folder (NinjaAdventure/Slide___001.png, etc...) while they should be referred to without this sub folder (Slide___001.png, etc...).
<img width="226" alt="Wrong filename" src="https://user-images.githubusercontent.com/1280130/103444360-fa52b980-4c67-11eb-971d-0bc419c4efa9.png">

> Is it only for when making a asset.json?
> Does the importer automatically ignore image files based on its existence?

Yes and yes. If you had a file called `IGNORED_FOR_IMAGE_ASSETS.md`, the importer won't try to automatically create object from the images found in this folder. It's only useful for your folder containing a `.asset.json` file

> The files I used had an extra underscore in them and don't contain the asset name, file names don't matter for asset.json right?

Yes, there is no convention for the `.asset.json` file, but as with others the filename should be in plain English:
`My Super Ninja.asset.json` (will be used to create an asset called `My Super Ninja`). The name of the resource files (the images) are not important because they won't be used to create objects... but still try to use similar name/same convention so that the creator that uses the asset does not have weird file names in its resources :) 

> Does GDevelop beta102 currently handle name collisions when importing assets? This is the second time I've run into this problem on this project.

Can you post the steps to reproduce this? Thanks!

**Comment by tristanbob** [comment]
In this video, I start from a new project and add two sprite objects with multiple frames each.   Source files for both objects have the same names, which GDevelop doesn't handle properly.

https://user-images.githubusercontent.com/8879811/103448259-34b25b80-4c54-11eb-9f96-3a0cd06e2d5b.mp4

**Comment by 4ian** [comment]
@tristanbob Ok I see, this is not related to the assets store so let's continue the discussion on a new issue :) 
There is no check for existing files made when you add a resource from another folder. The solution for now is to copy by yourself the images in your project folder.
The asset store does not have this issue as images are compared to existing resources :)

**Comment by Eliathx** [comment]
While working on the assets, there is this question I have:

![image](https://user-images.githubusercontent.com/67605537/104871927-67c64180-591a-11eb-991a-54de6131b38c.png)

In this one, they came as spritesheets, but since they are all the same size (32x32px) in Aseprite there is a quick way of having them ready in different frames that can later be saved as .png's and will have a number indexed for every frame. This sounds pretty quick. However, this might cause confusion as it may be seen as an animated object. But the problem here is: would we have to name every single icon/item with its name (like "IconGrape.png, IconWatermelon.png, ...")?

Notice that in this example there are 96 icons, all of them different. They could be counted and named as a group (let's say "RPGIcon") and then indexed, but I am not sure. 
On the other hand, naming 96 icons sounds very time consuming and make the process slow. Taking into account that there are **many** of these "collections", this would take quite a long time. Any suggestions/guidelines?

**Comment by Eliathx** [comment]
A second question, I might be a bit silly or missed something, but I didn't notice until I was working on them:

> "In the future, we'll support one _METADATA.json file for each animation. For now, just one is supported for all animations of the object."

Does this mean the frame rate in the .json applies to all animations in the object? If so, wouldn't it bring some problems, since not all animations have similar frame counts?

For example, in one object, some idle animations have about 4 to 6 frames and go at 100ms, but in the same object, they have a running, attacking, or any other animations with a higher frame count, meaning a different speed.

**Comment by 4ian** [comment]
Hey @Eliathx thanks for asking :) 
The short answer is: *yes and no* ;) I push for naming individual objects as such, but you can also use a generic name (see below)

> This sounds pretty quick. However, this might cause confusion as it may be seen as an animated object. But the problem here is: would we have to name every single icon/item with its name (like "IconGrape.png, IconWatermelon.png, ...")?

It's indeed pretty quick, but also this would not give a satisfying enough result for the users of the asset store, as the idea is to allow them to search directly by the name of the object or its tags.
It's true that depending on your game, if you're searching for an asset, you might want only some specific looking assets from a pack (to keep a visual coherency), so you'll probably filter the assets by this pack. But still, the idea is to things able to be searched by name :) 

Of course, this takes time, but not that much when you consider that you can open the folder with the PNGs in a file explorer/Finder, use the shortcut to rename the first file, look at what it is in the preview, type the name, press Enter, and go to the second file :) 
This will take you only a few minutes for a few dozen/hundred assets and will give a far superior search experience to people browsing the assets (especially if you put the assets in proper folders with `TAGS.md` files).

> Taking into account that there are many of these "collections", this would take quite a long time.

In this specific case, it might be ok to just take all the files and rename them to `Generic RPG icon (1).png`, `Generic RPG icon (2).png`...
You can also name them `Generic RPG icon 1.png`, `Generic RPG icon 2.png`, ... - just be sure NOT to use an underscore (`Generic RPG icon_1.png`) otherwise they would be considered as the same object with multiple frames.

Let me know if that sounds ok to you?
It's not as good as going ahead and renaming each file as "Key.png", "Map.png", "Blue gem.png" - but still fine enough!

> Does this mean the frame rate in the .json applies to all animations in the object? If so, wouldn't it bring some problems, since not all animations have similar frame counts?

Yes that's a limitation. But as I wrote "In the future, we'll support one _METADATA.json file for each animation", so I suggest that if you have an object like this, to this convention. For example:

* Idle animation: `Knight_Idle_METADATA.json` for the idle animation (with images: `Knight_Idle_1.png`, `Knight_Idle_2.png`...)
* Attack animation: `Knight_Attack_METADATA.json` for the attack animation (with images: `Knight_Attack_1.png`, `Knight_Attack_2.png`...)

Does that sound ok to you too? Let me know :) 
Trying to find a good balance between enough flexibility and good convention so that we don't have to author an entire software just to import assets :)

**Comment by 4ian** [comment]
I've edited my post about the rules to make these two things clear :)

**Comment by Bouh** [comment]
Some categories in the asset store are empty, there are no images nor objects.

They are empty in the asset store but are present in the extension icon selector in the extension properties.
And vise-versa.

"Search new behaviors" have empty categories too, such as Internet filter, and many others.

**Comment by Silver-Streak** [comment]
Just to add to this here, around the above message from @Bouh 

The IDE absolutely allows you to add SVG as sprite animation images right now. Not sure if it's intended or not.

It appears to render them via rasterization, so there's no huge benefits to the SVG itself, but they load and work in preview, too.

Unsure if this is intended by @4ian as I can't find this anywhere in the update history.

**Comment by Bouh** [comment]
In fact GDevelop take the texture file and give it to PixiJS, textures for a sprite can be in theory HTMLImageElement | HTMLCanvasElement | SVGElement | HTMLVideoElement.
Because PixiJS allow that.

SVG should not be supported in the resource tab, scan and remove useless SVG files shouldn't work.
And yes PixiJS pixelize the SVG in a buffer texture before rendering it.

**Comment by 4ian** [comment]
A bit off topic but yeah to confirm that 1) some categories are empty, I would need to remove them on the server side I think and 2) SVG seems to work because PixiJS does some magic under the hood to render them but for now I prefer not to officially support them, because it would need a commitment from us to support them more or less indefinitely in the future... and if we switch later to another internal rendering engine (Haxe, Cocos2d, whatever...) it's unsure a lot of them can do this. I would almost prefer something in GDevelop itself that would take your SVG and render them as pngs on the fly so that it's more portable.

**Comment by tristanbob** [comment]
I want to add some additional guidance on audio files, based on conversations and testing with @4ian. 

### Summary:
Convert your audio files to the AAC format using fre:ac, with a quality setting of "3".

### Details:
A great tool to convert audio files to AAC format is [fre:ac](https://www.freac.org/).  fre:ac is free, opensource, and cross-platform.  It can process batches of files, with many options on how they are named and put into folders.

- Choose the encoder called "Fraunhofer FDK AAC Encoder".
- Select "File format > AAC"
- Select "Quality > Set Quality > 3"

This will create high-quality audio files, but using a variable bit rate that won't exceed 96 kbps (assuming stereo input).  

Related Discussion:  https://github.com/4ian/GDevelop/issues/2378

![Screenshot 2021-03-11 112608](https://user-images.githubusercontent.com/8879811/110835883-9bf41980-825c-11eb-93c8-ef795b0379a2.jpg)

![Screenshot 2021-03-11 112527](https://user-images.githubusercontent.com/8879811/110835884-9c8cb000-825c-11eb-9b8d-f22228a71329.jpg)

**Comment by 4ian** [comment]
Thanks! This was added to the wiki page with the rules to submit to the assets store. I think we could almost close this issue and have conversations on the relevant card of the Trello board?

**Comment by Bouh** [comment]
When adding an object via the asset store there are empty lists because they refer to sounds, there is no sound object.
